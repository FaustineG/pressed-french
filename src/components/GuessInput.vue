<script setup lang="ts">
import { computed, ref } from 'vue'

const props = defineProps<{ modelValue: string; availableLetters: string[] }>()
const emits = defineEmits<{
  (e: 'update:modelValue', newValue: string): void
  (e: 'enter'): void
}>()

const letters = ref(props.availableLetters)
const allClicked = ref([false, false, false, false, false, false])

const lettersClicked = computed(() =>
  allClicked.value.map((clicked, index) => ({ letter: letters.value[index], clicked, index }))
)

// const isCorrectGuess = (guess: string) => {
//   let tempLetters = letters.value.toString()
//   return guess
//     .split('')
//     .map((letter) => {
//       const includesLetter = tempLetters.includes(letter)
//       if (includesLetter) {
//         tempLetters = tempLetters.replace(letter, '')
//       }
//       return includesLetter
//     })
//     .reduce((acc, val) => acc && val, true)
// }

const inputValue = computed({
  get: () => {
    return props.modelValue
  },
  set: (newValue) => {
    emits('update:modelValue', newValue)
  }
})

const onKeydown = (e: KeyboardEvent) => {
  if (e.key === 'Enter') {
    onEnter()
  } else if (e.key === 'Alt' || e.key === 'Control' || e.key === 'Shift') {
    shuffle()
  } else if (e.key === 'Backspace') {
    const indexLetter = letters.value.indexOf(inputValue.value.slice(-1))
    allClicked.value[indexLetter] = false
    inputValue.value = inputValue.value.slice(0, -1)

    return
  } else {
    // const letter = e.key.toLowerCase()
    const availableLetter = lettersClicked.value.find(
      (l) => l.letter === e.key && l.clicked === false
    )
    if (availableLetter) {
      inputValue.value += availableLetter.letter
      allClicked.value[availableLetter.index] = true
    }
  }
}

const clickLetter = (letter: string, index: number, clicked: boolean) => {
  if (clicked) {
    inputValue.value = inputValue.value.slice(0, -1)
    allClicked.value[index] = false
  } else {
    inputValue.value = inputValue.value.concat(letter)
    allClicked.value[index] = true
  }
}

const onEnter = () => {
  emits('enter')
  allClicked.value = [false, false, false, false, false, false]
}

const shuffle = () => {
  letters.value = letters.value.sort(() => 0.5 - Math.random())
  inputValue.value = ''
  allClicked.value = [false, false, false, false, false, false]
}

const onDelete = () => {
  const availableLetter = lettersClicked.value.find(
    (l) => l.letter === inputValue.value.slice(-1) && l.clicked === true
  )
  if (availableLetter) {
    inputValue.value = inputValue.value.slice(0, -1)
    allClicked.value[availableLetter?.index] = false
  }
}
</script>

<template>
  <section class="letter-section">
    <div class="letters">
      <button
        v-for="{ letter, index, clicked } in lettersClicked"
        :key="letter + index.toString()"
        @click="() => clickLetter(letter, index, clicked)"
        class="letter-button"
        :class="{ disabled: clicked }"
      >
        {{ letter }}
      </button>
    </div>
    <button @click="onDelete" class="delete-button">&lt;</button>
    <button @click="shuffle" class="shuffle-button"><img src="../assets/refresh.svg" /></button>
  </section>
  <section class="input-section">
    <input type="text" v-model="inputValue" @keydown.prevent="onKeydown" />
    <button @click="onEnter">ok</button>
  </section>
</template>

<style scoped lang="scss">
.letter-section {
  --letter-size: 3rem;
  --font-size: 2rem;
  display: flex;
  margin-bottom: 1rem;
  gap: 2rem;

  .letters {
    display: flex;
    .letter-button {
      border-radius: 0;
      width: var(--letter-size);
      height: var(--letter-size);
      font-size: var(--font-size);
      margin: 0 0.5rem;
      font-variant: small-caps;
    }

    .disabled {
      background-color: grey;
    }
  }

  .shuffle-button,
  .delete-button {
    width: var(--letter-size);
    height: var(--letter-size);
    font-size: var(--font-size);
    display: flex;
    align-items: center;
    justify-content: center;
  }
}

@media (max-width: 600px) {
  * {
    --letter-size: 5rem;
    --font-size: 4rem;
  }
}

.input-section {
  min-height: 3rem;

  display: flex;
  input {
    border-radius: 5px 0 0 5px;
    font-size: var(--font-size);
    border: none;
  }

  button {
    border-radius: 0 5px 5px 0;
  }
}
</style>
