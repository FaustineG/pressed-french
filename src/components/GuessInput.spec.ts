import { describe, expect, it } from 'vitest'
import GuessInput from './GuessInput.vue'
import { mount } from '@vue/test-utils'

const initComponent = (propsData = {}) => {
  return mount(GuessInput, {
    props: {
      modelValue: '',
      availableLetters: [],
      ...propsData
    }
  })
}

describe('GuessInput', () => {
  it('should match snapshot', () => {
    const wrapper = initComponent()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to click on letter', () => {
    const wrapper = initComponent()
    const allLetterButtonElements = wrapper.findAll('.letter-button')
  })
})
