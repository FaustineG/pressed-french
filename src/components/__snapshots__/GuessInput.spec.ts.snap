// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GuessInput > should display all 1`] = `
"<section data-v-550e92e3="" class="letter-section">
  <div data-v-550e92e3="" class="letters"><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button></div>
</section>
<section data-v-550e92e3="" class="buttons-section"><button data-v-550e92e3="" class="delete-button">&lt;</button><button data-v-550e92e3="" class="shuffle-button"><img data-v-550e92e3="" src="/src/assets/refresh.svg"></button></section>
<section data-v-550e92e3="" class="input-section"><input data-v-550e92e3="" type="text"><button data-v-550e92e3="">ok</button></section>"
`;

exports[`GuessInput > should match snapshot 1`] = `
"<section data-v-550e92e3="" class="letter-section">
  <div data-v-550e92e3="" class="letters"><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button><button data-v-550e92e3="" class="letter-button"></button></div>
</section>
<section data-v-550e92e3="" class="buttons-section"><button data-v-550e92e3="" class="delete-button">&lt;</button><button data-v-550e92e3="" class="shuffle-button"><img data-v-550e92e3="" src="/src/assets/refresh.svg"></button></section>
<section data-v-550e92e3="" class="input-section"><input data-v-550e92e3="" type="text"><button data-v-550e92e3="">ok</button></section>"
`;
